allow to get Invoicepayments from Invoice

julienfastre opened the pull request:
This PR add the ability to retrieve `InvoicePayements` from `Invoice` object.

It adds a method `getPayments($withPluginInfo = false, $withAttempts = false, $headers = null)` on `Invoice` class.

julienfastre commented:
Note that there isn't any test associated with this feature. Could you provide some guidance to add test on this ?

pierre commented on src/InvoicePayment.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+/*
+ * Copyright 2011-2017 Ning, Inc.
```
Copyright should be attributed to `The Billing Project, LLC` instead.

julienfastre replied:
ok.

pierre commented on src/InvoicePayment.php:
```diff
@@ -0,0 +1,29 @@
+/**
+ * 
+ *
+ * @author Julien Fastré <julien.fastre@champs-libres.coop>
```
We don't track individual contributors in source files (authorship will remain in Git / GitHub though!).

julienfastre replied:
Ok. This is auto-generated stuff, :-)

sbrossie commented on src/Invoice.php:
```diff
@@ -57,6 +57,50 @@ public function get($withItems, $headers = null)
+    
+    /**
+     * Get payements for this invoice
+     * 
```
Nit: typo `payements` -> `payments `

julienfastre commented:
Thanks. I will fix the phpcs warnings.

I am also currently working on the counterpart `createPayment` method. and I created an issue (#24) for discussing this feature. If this is accepted, I will update this PR. (Or should I create another one ?)

MaxGfeller commented:
Looks good to me. Regarding the tests see [this comment](https://github.com/killbill/killbill-client-php/pull/15#issue-220219085) too.

pierre commented:
> If this is accepted, I will update this PR. (Or should I create another one ?)

A separate PR would be easier to review. Thanks!

julienfastre commented:
I fixed the issues with code formatting, and added some tests. The tests pass with a real KB server (`composer test-integration`) but not with the mocked server.

Without my test, I get plenty of errors in tests: there are missing `test/resources/mock` files (examples : `test/resources/mocks/001a9760ef92b47292cfb3a356884005`). Is that normal ? But I think that no, because other test passes (https://travis-ci.org/killbill/killbill-client-php/builds/237432323).

GregOriol commented on src/InvoicePayment.php:
```diff
@@ -0,0 +1,27 @@
+ */
+class InvoicePayment extends InvoicePaymentAttributes
+{
+    const SUB_PATH_INVOICES_PAYMENT = '/payments';
```
Currently all paths are declared in the Client class, wouldn't it be more coherent to put this one there too? (ie. also instead of creating a mostly empty file)

julienfastre replied:
ok.

GregOriol commented on src/Invoice.php:
```diff
@@ -58,6 +60,46 @@ public function get($withItems, $headers = null)
+     * @param  boolean       $withPluginInfo
+     * @param  boolean       $withAttempts
+     * @param  string[]|null $headers        Any additional headers
+     * @return InvoicePayment[]
```
There should be a spacing line between "@param"s and the "@return".

Also, the "Any additional headers" should be one more space away to the right, because it has to be column aligned with the end of "Info" above.

julienfastre replied:
ok. I followed the space requirements given by phpcs.

julienfastre replied:
@GregOriol 

> Also, the "Any additional headers" should be one more space away to the right, because it has to be column aligned with the end of "Info" above.

Sorry, PHPCS complaints about adding an extra space before "Any additionnal headers" : https://travis-ci.org/killbill/killbill-client-php/jobs/306525628

GregOriol replied:
PHPCS is probably right then :-)

julienfastre commented:
>  If you added tests (and calls to the killbill APIs in them), you have to generate the mocks for them, so that it is possible to run the integration tests without a real killbill instance (with travis-ci for ex). Check the "record-mocks" script in composer.json to do so.

Ok. I think that was the tool I needed.

julienfastre commented:
@GregOriol 

I ran `` and I get a lot of new files: 

```
	modifié :         test/resources/mocks/28355025c28be7bfe237e20a042926eb
	modifié :         test/resources/mocks/6b23c2a7c56790d96163b5364e6a566c
	modifié :         test/resources/mocks/890df175c8b9765b26d5eae5d71a5403
	modifié :         test/resources/mocks/b5fda36a4f1eb76453f0c917d445bb4d
	modifié :         test/resources/mocks/c2acae236166d219e616b5e3349d344b
	modifié :         test/resources/mocks/externalKey-0bd7be995dbdb6281bc61821db7a75a4
	modifié :         test/resources/mocks/externalKey-0dad489c6d35c6c56abe8050749b1683
	modifié :         test/resources/mocks/externalKey-14d5b2899761d82e965dbac62674609f
	modifié :         test/resources/mocks/externalKey-1a53b1621549748f443d8b3f422ab094
	modifié :         test/resources/mocks/externalKey-2404841bf9f6a8b3d4a978b364471ec8
	modifié :         test/resources/mocks/externalKey-26e401bc5d24dda56d665ab024a2b008
	modifié :         test/resources/mocks/externalKey-40e6899680455ebc9121d6e3ac6347d6
	modifié :         test/resources/mocks/externalKey-657d16acb5d47e73a39bf71d8c2973b0
	modifié :         test/resources/mocks/externalKey-68efceced4cec7b4307e35f3b3455786
	modifié :         test/resources/mocks/externalKey-6bf46fd483ca397fc2cfd70ea6976ae7
	modifié :         test/resources/mocks/externalKey-7901263a72bfe16f2f7fa45bf88440a3
	modifié :         test/resources/mocks/externalKey-799609342b7e1f425f673fd0435bca2a
	modifié :         test/resources/mocks/externalKey-80d45ca64525042aff315ee593737a5d
	modifié :         test/resources/mocks/externalKey-80dcfa75531f7727d4813c271c388bd7
	modifié :         test/resources/mocks/externalKey-8a80e8e6a27956c82f8e1fbfa2840c23
	modifié :         test/resources/mocks/externalKey-8aad4e101c66e65febce3d2166ce74ba
	modifié :         test/resources/mocks/externalKey-8bb4d73077ebf2315e4ba9cd2503d7fb
	modifié :         test/resources/mocks/externalKey-a5ec57f633402a659d2d9c8706fd8d34
	modifié :         test/resources/mocks/externalKey-a62f1f3b750b9537c18a7fcb4fb7b461
	modifié :         test/resources/mocks/externalKey-af8fd623e1bc460b9193bbec9f194305
	modifié :         test/resources/mocks/externalKey-b1a7cccaff5bbd3703ce080b2b15419a
	modifié :         test/resources/mocks/externalKey-bc57ecec1f7e51cec30dbb50075273e6
	modifié :         test/resources/mocks/externalKey-c71780840cfaa412ed9dd82694dedc31
	modifié :         test/resources/mocks/externalKey-cfab791689d0048e88a5dba886036bf1
	modifié :         test/resources/mocks/externalKey-d5b5e224113c4df1122367acbfabd894
	modifié :         test/resources/mocks/externalKey-e344d571f9c30286bde4b0cdfa860795
	modifié :         test/resources/mocks/externalKey-ec09e92d83ce15cea36fda8dfb8702d9
	modifié :         test/resources/mocks/externalKey-f4db55a641340e59aaac0c101e2ff263
	modifié :         test/resources/mocks/f8137d21685e7b29a02eb74159764138

Fichiers non suivis:
  (utilisez "git add <fichier>..." pour inclure dans ce qui sera validé)

	src/Credit.php
	test/resources/mocks/011e1e9b436b4fabfe111e786211b6ba
	test/resources/mocks/0154746a1c5142ceaaa8d1ddc6bb9b69
	test/resources/mocks/02f50fd8577e4af4c0731a6b0c6470a3
# truncated, there are about 50-100 files
```

should I commit all of them ?

GregOriol commented:
@julienfastre you should actually delete all the files in test/resources/mocks/, then record yours and commit all of those (mocks have some unique ids in them so the add/delete/modify thing doesn't make much sense)

julienfastre commented:
I am quite disappointed: the test pass locally, but are not correctly executed by travis (https://travis-ci.org/killbill/killbill-client-php/jobs/306527029)...

julienfastre commented:
oups... I made a mistake an pushed a lot of modifications I should separate on different PR. I will close this PR and will be back in January about this... sorry.
